gctags/gctags.1
.\"
.\" Copyright (c) 1996, 1997, 1998, 1999
.\"             Shigio Yamaguchi. All rights reserved.
.\" Copyright (c) 1999, 2000
.\"             Tama Communications Corporation. All rights reserved.
.\"
.\" This file is part of GNU GLOBAL.
.\"
.\" GNU GLOBAL is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2, or (at your option)
.\" any later version.
.\"
.\" GNU GLOBAL is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program; if not, write to the Free Software
.\" Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
.\"
.Dd Oct 10, 1999
.Dt GCTAGS 1
.Os FreeBSD
.Sh NAME
.Nm gctags
.Nd print cross reference list for gtags.
.Sh SYNOPSIS
.Nm gctags
.Op Fl b
.Op Fl e
.Op Fl n
.Op Fl r
.Op Fl s
.Op Fl w
.Ar file ...
.Sh DESCRIPTION
.Nm Gctags
print cross reference list for
.Xr gtags 1
from the specified C,
.Tn C++ ,
.Tn YACC ,
.Tn JAVA ,
and assembler source to standard output.
Each line of output contains the object name, the line number which it appears,
the file in which it is defined, and a line image separated by
white-space.
It's same with the output of
.Xr ctags 1
with -x option.
.Pp
Depending upon the options provided to
.Nm gctags ,
objects will consist of function definitions, function references and other
symbols.
This command is the default parser of GLOBAL source code tag system.
.Pp
.Bl -tag -width Ds -compact
.It Fl b
.It Fl -begin-block
Force level 1 block to begin when reach a '{' at the first column. (C only)
.It Fl e
.It Fl -end-block
Force level 1 block to end when reach a '}' at the first column. (C only)
.It Fl n
.It Fl -no-tags
Suppress output of tags. It is useful to use with
.Fl w
option.
.It Fl r
.It Fl -reference
Locate function references instead of function definitions. GTAGS file is
needed at the current directory. (C, C++ and Java source only)
By default, locate function definitions.
.It Fl s
.It Fl -symbol
Collect symbols other than functions. By default, locate function definitions.
.It Fl w
.It Fl -warning
Print warning message.
.El
.Pp
The
.Fl r
and
.Fl s
options override each other; the last one specified determines the method used.
.Pp
Files whose names end in
.Nm \&.c
or
.Nm \&.h
are assumed to be C
source files and are searched for C style routine and macro definitions.
Files whose names end in
.Nm \&.c++
.Nm \&.cc
.Nm \&.cpp
.Nm \&.cxx
.Nm \&.hxx
.Nm \&.C
.Nm \&.H
are assumed to be C++
source files.
Files whose names end in
.Nm \&.y
are assumed to be
.Tn YACC
source files.
Files whose names end in
.Nm \&.java
are assumed to be Java source files.
Files whose names end in
.Nm \&.s
or
.Nm \&.S
are assumed to be Assembler
source files.  Other files are searched for C style definitions.
.Pp
Yacc files each have a special tag.
.Ar Yyparse
is the start
of the second section of the yacc file.
.Sh DIAGNOSTICS
.Nm Gctags
exits with a non 0 value if an error occurred, 0 otherwise.
Duplicate objects are not considered errors.
.Sh SEE ALSO
.Xr btreeop 1 ,
.Xr ex 1 ,
.Xr global 1 ,
.Xr gtags 1 ,
.Xr htags 1 ,
.Xr vi 1 .
.Sh BUGS
.Pp
.Nm Gctags
relies on the input being well formed, and any syntactical
errors will completely confuse it.
.Pp
Assembler support is far from complete.  It extracts only ENTRY()
and ALTENTRY() from source file. Probably valid only for FreeBSD and Linux
kernel source.
.Sh AUTHORS
Shigio Yamaguchi
.Sh HISTORY
The
.Nm
command appeared in FreeBSD 2.2.2.


NEWS
___________________________________
|      |  |  |     |  _  |     |  |
|  |___|  |  |  |  |    _|  |  |  |    GNU GLOBAL source code tagging system
|  |   |  |  |  |  |     |     |  |
|  ~~  |   ~~|     |  ~  |  |  |   ~~|          for all hackers.
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 Copyright (c) 2000-2016 Tama Communications Corporation

 This file is free software; as a special exception the author gives
 unlimited permission to copy and/or distribute it, with or without
 modifications, as long as this notice is preserved.

 This program is distributed in the hope that it will be useful, but
 WITHOUT ANY WARRANTY, to the extent permitted by law; without even the
 implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

		----------------------------------

GNU GLOBAL NEWS - User visible changes.
At least, you should search '[INCOMPATIBLE CHANGES]' for each version.

Version 6.5.7 - 15 May 2017

[CHANGES]
o global: Now global(1) accept a file as --nearness option.
  If it is a file, tags in the file are shown first.
o global: Added aliases for MAKEOBJDIR and MAKEOBJDIRPREFIX.
  GTAGSOBJDIR: alias for MAKEOBJDIR
  GTAGSOBJDIRPREFIX: alias for MAKEOBJDIRPREFIX
  These aliases are given higher priority than the originals.
o global: New command --print=<name>.
  <name> may be one of: root, dbpath or conf.
  This is a generic version of the -p (--print-dbpath) command.
	'global -p' is equal to 'global --print=dbpath'
	'global -pr' is equal to 'global --print=root'
o gtags: Updated the built-in PHP parser to support namespaces and traits,
  which were introduced in PHP5. The following keywords have been removed:
	- failure: not a keyword
	- header: normal built-in function
	- is_array: normal built-in function
	- is_set: typo -> isset
	- printf: normal built-in function
	- setcookie: normal built-in function
	- success: not a keyword

[FIXED BUGS]
o globash: Echo command with -n option didn't work on some OS. Now it works.
o global: The --nearness option without -x option brought segmentation fault.
	$ global --nearness=. main
	Segmentation fault: 11
  Now it works.
o global: --from-here with -i option did not ignore case of function name
  in the context. Now it works.
	Input:
	[main.c]
	1: main() {
	2:         func();
	3:         Func();
	4: }
	5: func() {}
	6: Func() {}

	$ global -xi --from-here=6:main.c func
	(global-6.5.6)
	Func                6 main.c           Func() {}
	func                5 main.c           func() {}
	(global-6.5.7)
	Func                3 main.c            Func();
	func                2 main.c            func();

[INCOMPATIBLE CHANGES]
o gtags.conf: Added new candidate path.
        (1) --gtagsconf file
        (2) $GTAGSCONF
        (3) [project root]/gtags.conf
        (4) [project root]/[objdir]/gtags.conf       <- [NEW]
        (5) $HOME/.globalrc
        (6) /etc/gtags.conf
        (7) [sysconfdir]/gtags.conf

        The default of [objdir] is 'obj'. You can change it by
        the environment variables (GTAGSOBJDIR or MAKEOBJDIR).

Version 6.5.6 - December 19 2016

[CHANGES]
New facilities:
o htags-server: New --retry[=n] option.
  If the port is already in use, retry n times with incrementing the port number.
  The default of n is 20.
o htags: Changed the format of function header (--show-position) to make copying
  text easier.
o geco.rc: Added 'fzf' as a selector candidate.
o gtags: New configuration variable 'gtags_hook'.
  gtags_hook=<command line>
	Specify a command line which should be executed at the beginning of gtags(1).
	Leading "./" in any path is always means the project root directory, since
	gtags(1) is always invoked there.
  GTAGS_COMMANDLINE environment variable (read only)
	You can get the effective command line of gtags(1) from the hook. It includes
	both $GTAGS_OPTIONS and real arguments.

   [Usage]
   You can update 'gtags.files' before gtags(1) read it.

	[gtags.conf]
	+----------------------------------------
	|...
	|:gtags_hook=find src lib -print >gtags.files:

   You can refer the effective arguments of gtags(1) from the hook using
   environment variable GTAGS_COMMANDLINE.

	[gtags.conf]
	+----------------------------------------
	|...
	|:gtags_hook=./gen.sh:\
	|:GTAGS_OPTIONS=-c:\

	[gen.sh]
	+----------------------------------------
	|#!/bin/sh
	|echo ">>> $GTAGS_COMMANDLINE"		# show effective command line

	$ gtags -O
	>>> gtags -c -O
	$ _

[INCOMPATIBLE CHANGES]
o htags: Now --cvsweb option always insert 'view=log' to the generated URLs.
  Because it seems to be almost always a desirable specification.

[FIXED BUGS]
o gtags: Gtags often aborts with a message "buffer overflow. strlimcpy" when
  it encounters a long token (> 152 bytes). Now gtags always ignores it with
  a message "symbol name is too long. (Ignored)".
o htags: Old packages included two CGI scripts (completion.cgi, global.cgi)
  generated in the release manager's machine by mistake. They have some literal
  path like '/opt/local/bin/perl' which works only with MacPorts.
  Now, they are generated in the target (your) machine.

Version 6.5.5 - Sep 21 2016

This is a bug fix release.

[FIXED BUGS]
o exuberant-ctags.la: The path name of ctags command couldn't changed.
  Now, it can be changed by configuration variable 'ctagscom'.
o Tag files got corrupted when the built-in C++ parser gave up parsing and aborted.
  Now, it is avoided.

o Some inappropriate error messages were changed.

  Error1: NULL file (file size == 0)

	$ cp /dev/null GTAGS
	$ global -x main

	(GLOBAL-6.5.4)	global: /tmp/test/GTAGS seems older format. Please remake tag files.
	(GLOBAL-6.5.5)	global: /tmp/test/GTAGS seems to be corrupted.

  Error2: Corrupted file

	$ echo aaa >GTAGS
	$ global -x main

	(GLOBAL-6.5.4)	global: GTAGS not found.
	(GLOBAL-6.5.5)	global: /tmp/test/GTAGS seems to be corrupted.

Version 6.5.4 - Mar 27 2016

This is a bug fix release.

[CHANGES]
Changed:
o gtags.conf: The entries which include '.xml' were commented out,

gctags/gctags.c
/*
 * Copyright (c) 1998, 1999 Shigio Yamaguchi
 * Copyright (c) 1999, 2000, 2001, 2002 Tama Communications Corporation
 *
 * This file is part of GNU GLOBAL.
 *
 * GNU GLOBAL is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2, or (at your option)
 * any later version.
 *
 * GNU GLOBAL is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 */

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif
#ifdef HAVE_LIMITS_H
#include <limits.h>
#endif
#include <stdio.h>
#ifdef STDC_HEADERS
#include <stdlib.h>
#endif
#ifdef HAVE_STRING_H
#include <string.h>
#else
#include <strings.h>
#endif
#ifdef HAVE_UNISTD_H
#include <unistd.h>
#endif
#include "getopt.h"

#include "global.h"
#include "gctags.h"
#include "const.h"

char	*notfunction;

int	main(int, char **);
static	void usage(void);
static	void help(void);
static	int match_suffix_list(const char *, const char *);

struct words *words;
static int tablesize;

/*
 * language map.
 *
 * By default, default_map is used.
 */
static const char *default_map = "c:.c.h,yacc:.y,asm:.s.S,java:.java,cpp:.c++.cc.cpp.cxx.hxx.C.H";
static char *langmap;
static STRBUF *active_map;

int	bflag;			/* -b: force level 1 block start */
int	dflag;			/* -d: treat #define with no argument */
int	eflag;			/* -e: force level 1 block end */
int	nflag;			/* -n: doen't print tag */
int	rflag;			/* -r: function reference */
int	sflag;			/* -s: collect symbols */
int	tflag;			/* -t: treat typedefs, structs, unions, and enums. */
int	wflag;			/* -w: warning message */
int	vflag;			/* -v: verbose mode */
int	do_check;
int	show_version;
int	show_help;
int	debug;

int	yaccfile;		/* yacc file */

static void
usage()
{
	fputs(usage_const, stderr);
	exit(2);
}
static void
help()
{
	fputs(usage_const, stdout);
	fputs(help_const, stdout);
	exit(0);
}
static struct option const long_options[] = {
	{"begin-block", no_argument, NULL, 'b'},
	{"check", no_argument, &do_check, 1},
	{"define", no_argument, NULL, 'd'},
	{"end-block", no_argument, NULL, 'e'},
	{"no-tags", no_argument, NULL, 'n'},
	{"reference", no_argument, NULL, 'r'},
	{"symbol", no_argument, NULL, 's'},
	{"typedef", no_argument, NULL, 't'},
	{"verbose", no_argument, NULL, 'v'},
	{"warning", no_argument, NULL, 'w'},

	/* long name only */
	{"debug", no_argument, &debug, 1},
	{"langmap", required_argument, NULL, 0},
	{"version", no_argument, &show_version, 1},
	{"help", no_argument, &show_help, 1},
	{ 0 }
};

int
main(argc, argv)
	int	argc;
	char	**argv;
{
	char	*p;
	int	flag;
	int	optchar;
	int     option_index = 0;

	while ((optchar = getopt_long(argc, argv, "bdenrstvw", long_options, &option_index)) != EOF) {
		switch(optchar) {
		case 0:
			p = (char *)long_options[option_index].name;
			if (!strcmp(p, "langmap"))
				langmap = optarg;	
			break;
		case 'b':
			bflag++;
			break;
		case 'd':
			dflag++;
			break;
		case 'e':
			eflag++;
			break;
		case 'n':
			nflag++;
			break;
		case 'r':
			rflag++;
			sflag = 0;
			break;
		case 's':
			sflag++;
			rflag = 0;
			break;
		case 't':
			tflag++;
			break;
		case 'v':
			vflag++;
			break;
		case 'w':
			wflag++;
			break;
		default:
			usage();
			/* NOTREACHED */
		}
	}
	if (show_version)
		version(NULL, vflag);
	else if (show_help)
		help();
	else if (do_check) {
		fprintf(stdout, "Part of GLOBAL\n");
		exit(0);
	}
	/*
	 * construct language map.
	 */
	active_map = strbuf_open(0);
	strbuf_puts(active_map, (langmap) ? langmap : default_map);
	flag = 0;
	for (p = strbuf_value(active_map); *p; p++) {
		/*
		 * "c:.c.h,java:.java,cpp:.C.H"
		 */
		if (flag == 0 && *p == ',' || flag == 1 && *p == ':')
			die_with_code(2, "syntax error in --langmap option.");
		if (*p == ':' || *p == ',') {
			flag ^= 1;
			*p = 0;
		}
	}
	if (flag == 0)
		die_with_code(2, "syntax error in --langmap option.");

        argc -= optind;
        argv += optind;

	if (argc < 1)
		usage();
	if (getenv("GTAGSWARNING"))
		wflag++;	
	/*
	 * This is a hack for FreeBSD.
	 * In the near future, it will be removed.
	 */
#ifdef __DGJPP__
	if (test("r", NOTFUNCTION) || test("r", DOS_NOTFUNCTION))
#else
	if (test("r", NOTFUNCTION))
#endif
	{
		FILE	*ip;
		STRBUF	*sb = strbuf_open(0);
		STRBUF	*ib = strbuf_open(0);
		char	*p;
		int	i;

		if ((ip = fopen(NOTFUNCTION, "r")) == 0)
#ifdef __DJGPP__
			if ((ip = fopen(DOS_NOTFUNCTION, "r")) == 0)
#endif
			die("'%s' cannot read.", NOTFUNCTION);
		for (tablesize = 0; (p = strbuf_fgets(ib, ip, STRBUF_NOCRLF)) != NULL; tablesize++)
			strbuf_puts0(sb, p);
		fclose(ip);
		if ((words = (struct words *)malloc(sizeof(struct words) * tablesize)) == NULL)
			die("short of memory.");
		/*
		 * Don't free *p.
		 */
		p = (char *)malloc(strbuf_getlen(sb) + 1);
		memcpy(p, strbuf_value(sb), strbuf_getlen(sb) + 1);
		for (i = 0; i < tablesize; i++) {
			words[i].name = p;
			p += strlen(p) + 1;
		}
		qsort(words, tablesize, sizeof(struct words), cmp);
		strbuf_close(sb);
		strbuf_close(ib);
	}

	/*
	 * pick up files and parse them.
	 */
	for (; argc > 0; argv++, argc--) {
		char *lang, *suffix, *list, *tail;

		/* initialize token parser. */
		if (!opentoken(argv[0]))
			die("'%s' cannot open.", argv[0]);
#if defined(_WIN32) || defined(__DJGPP__)
		/* Lower case the file name since names are case insensitive */
		strlwr(argv[0]);
#endif
		/* get suffix of the path. */
		suffix = locatestring(argv[0], ".", MATCH_LAST);
		if (!suffix)
			continue;
		list = strbuf_value(active_map);
		tail = list + strbuf_getlen(active_map);

		/* check whether or not list includes suffix. */
		while (list < tail) {
			lang = list;
			list = lang + strlen(lang) + 1;
			if (match_suffix_list(suffix, list))
				break;
			lang = NULL;
			list += strlen(list) + 1;
		}
		if (lang == NULL)
			continue;
		if (vflag)
			fprintf(stderr, "suffix '%s' assumed language '%s'.\n", suffix, lang);
		/*
		 * call language specific parser.
		 */
		if (!strcmp(lang, "c")) {
			if (!locatestring(suffix, ".h", MATCH_COMPLETE) && isCpp())
				Cpp();
			else
				C(0);
		} else if (!strcmp(lang, "yacc")) {
			C(YACC);
		} else if (!strcmp(lang, "asm")) {
			assembler();
		} else if (!strcmp(lang, "java")) {
			java();
		} else if (!strcmp(lang, "cpp")) {
			Cpp();
		}
		closetoken();
	}
	return 0;
}

/*
 * return true if suffix matches with one in suffix list.
 */
int
match_suffix_list(suffix, list)
	const char *suffix;
	const char *list;
{
	while (*list) {
		if (locatestring(list, suffix, MATCH_AT_FIRST))
			return 1;
		for (list++; *list && *list != '.'; list++)
			;
	}
	return 0;
}

int
cmp(s1, s2)
	const void *s1, *s2;
{
	return strcmp(((struct words *)s1)->name, ((struct words *)s2)->name);
}

int
isnotfunction(name)
	char	*name;
{
	struct words tmp;
	struct words *result;

	if (words == NULL)
		return 0;
	tmp.name = name;
	result = (struct words *)bsearch(&tmp, words, tablesize, sizeof(struct words), cmp);
	return (result != NULL) ? 1 : 0;
}

void
dbg_print(level, s)
	int level;
	const char *s;
{
	if (!debug)
		return;
	fprintf(stderr, "[%04d]", lineno);
	for (; level > 0; level--)
		fprintf(stderr, "    ");
	fprintf(stderr, "%s\n", s);
}
